File: 25. Reverse Nodes in k-Group/solution_iterative.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        dummy = ListNode(0)
        dummy.next = head
        newHead = False
        preStart = dummy
        while preStart is not None:
            tail = preStart
            for i in range(k):
                tail = tail.next
                if tail is None:
                    return dummy.next

            node, nextPreStart = preStart.next, preStart.next
            preStart.next = tail.next
            for i in range(k):
                nextNode = node.next
                node.next = preStart.next
                preStart.next = node
                node = nextNode
            if not newHead:
                dummy.next = preStart.next
                newHead = True
            preStart = nextPreStart
        return dummy.next

File: 25. Reverse Nodes in k-Group/solution_iterative.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseKGroup(head *ListNode, k int) *ListNode {
	dummy, newHead := &ListNode{Next: head}, false
	for preStart, tail := dummy, dummy; tail != nil; {
		for i := 0; i < k && tail != nil; i++ {
			tail = tail.Next
		}
		if tail == nil {
			break
		}
		prev := tail.Next
		for i, node := 0, preStart.Next; i < k; i++ {
			node.Next, prev, node = prev, node, node.Next
		}
		preStart.Next, preStart, tail = prev, preStart.Next, preStart.Next
		if !newHead {
			dummy.Next, newHead = prev, true
		}
	}
	return dummy.Next
}
